HRTB variants are missing a std::fmt::Pointer impl

It seems that there is a missing implementation of `std::fmt::Pointer` on some function pointers:

function pointers like `fn test() -> f32` can be passed as argument to `println!("{:p}", ...)`
but function pointers like `fn test(Vec<T> a) -> f32` can't because it's a HRTB variant.

see https://users.rust-lang.org/t/display-function-pointer/17073 for more details.


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

Triage: no change. This program: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=844d632aee1147f2dc9e7d05d3ef9462

```
fn show(a: fn () -> f32, b: fn (&Vec<i8>)->f32) {
    println!("the two pointers: {:p} {:p}", a, b);
}
```

gives this error:

```text
error[E0277]: the trait bound `for<'r> fn(&'r std::vec::Vec<i8>) -> f32: std::fmt::Pointer` is not satisfied
 --> src/lib.rs:2:48
  |
2 |     println!("the two pointers: {:p} {:p}", a, b);
  |                                                ^ the trait `std::fmt::Pointer` is not implemented for `for<'r> fn(&'r std::vec::Vec<i8>) -> f32`
  |
  = help: the following implementations were found:
            <extern "C" fn(A) -> Ret as std::fmt::Pointer>
            <extern "C" fn(A, ...) -> Ret as std::fmt::Pointer>
            <fn(A) -> Ret as std::fmt::Pointer>
            <unsafe extern "C" fn(A) -> Ret as std::fmt::Pointer>
          and 2 others
  = note: required by `std::fmt::Pointer::fmt`
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/library/coretests/tests/fmt/mod.rs
+++ b/library/coretests/tests/fmt/mod.rs
@@ -30,6 +30,12 @@ fn test_format_flags() {
     assert_eq!(format!("{:p} {:x}", p, 16), format!("{p:p} 10"));
 
     assert_eq!(format!("{: >3}", 'a'), "  a");
+
+    /// Regression test for <https://github.com/rust-lang/rust/issues/50280#issuecomment-626035934>.
+    fn show(a: fn() -> f32, b: fn(&Vec<i8>) -> f32) {
+        println!("the two pointers: {:p} {:p}", a, b);
+    }
+    show(|| 1.0, |_| 2.0);
 }
 
 #[test]

